api/agent/protocol/json.go

```diff
@@ -35,59 +63,125 @@ func writeString(err error, dst io.Writer, str string) error {
 
 // TODO(xxx): headers, query parameters, body - what else should we add to func's payload?
 // TODO(xxx): get rid of request body buffering somehow
-func (h *JSONProtocol) DumpJSON(req *http.Request) error {
+// @treeder: I don't know why we don't just JSON marshal this, this is rough...
```

Comment:
because if request body, we still don't have a better idea how to avoid buffering and stream it right into a container.

Reply:
Does it matter?  If it's too big, it'll blow up inside the container, either way it's getting fully loaded into an object. I'd vote for readable code/maintainability over a fraction of a millisecond of time marshaling to JSON. I believe you did this before and Reed had you change it ya?  I'll leave it for now, but as we add more fields, this will get harder and harder to maintain/read than it already is. 

And if we do want to support super large bodies, we should (must?) implement another scheme for it, eg: https://github.com/fnproject/fn/issues/500

Reply:
we have the ability to limit the size of the reader itself, that code is checked in as a server option; not really the issue. memory pressure / fragmentation is the real issue, json uses sync.Pool but it's just unnecessary overhead and those buffers are not properly sized and need expansion frequently, so it's really not doing such a great job across heterogeneous bodies. 

it would be nice if the json package had support for `MarshalTo(io.Writer) error` like normal writers, but it don't.

i can think of clever hacks to get mostly `json.Marshal` functionality, where we can put the fields in a temporary struct, marshal that, bite off the closing `}` and then copy that followed by `"body":""}` which may make the json dispatch slightly more palatable, but it seems the right place to optimize really.